### ISSUE TITLE
[1.8.1/1.9] Docker containers forget how to communicate with network

### ISSUE THREAD

**Issue by BenHall** [docker/docker#15788]
When launching a container it will not be able to connect with other containers or outside network. Not sure if this is related to the #15341 error. I've had the same issue with 1.8.1 and 1.9.

Very hard to reproduce and not sure what to look for when it happens. Docker inspect shows they have the IP address assigned.

```
$ uname -a
Linux host 3.13.0-61-generic #100-Ubuntu SMP Wed Jul 29 11:21:34 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

```
# docker version
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   bba762b
 Built:        Sun Aug 16 09:11:29 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   bba762b
 Built:        Sun Aug 16 09:11:29 UTC 2015
 OS/Arch:      linux/amd64
```

```
# docker info
Containers: 1048
Images: 215
Storage Driver: aufs
 Root Dir: /home/docker/data/aufs
 Backing Filesystem: extfs
 Dirs: 2313
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-61-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 8
Total Memory: 94.41 GiB
Debug mode (server): true
File Descriptors: 70
Goroutines: 140
System Time: 2015-08-24T11:51:39.356058733+02:00
EventsListeners: 1
Init SHA1:
Init Path: /usr/bin/docker
Docker Root Dir: /home/docker/data
```

```
root@host:~# docker run -it ubuntu bash
Error response from daemon: Cannot start container 68080554605abea4e9cf83632ac130a8c0588136a43a5c40755ef759eaf946a9: could not set link up for host interface vethbe5afcf: no such device
root@host:~# docker run -it ubuntu bash
root@7a8e403ffaea:/# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=58 time=4.15 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=58 time=4.02 ms
^C
--- 8.8.8.8 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 4.020/4.089/4.159/0.094 ms
root@7a8e403ffaea:/# exit
exit
root@host:~# docker run --rm -it ubuntu bash
root@880cc1d230ee:/# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
From 172.17.1.225 icmp_seq=1 Destination Host Unreachable
From 172.17.1.225 icmp_seq=2 Destination Host Unreachable
From 172.17.1.225 icmp_seq=3 Destination Host Unreachable
^C
--- 8.8.8.8 ping statistics ---
6 packets transmitted, 0 received, +3 errors, 100% packet loss, time 4999ms
```

**Comment by calvcoll** [comment]
Weirdly I've been having the same problem with all my containers on my local machine.

```
$ uname -a
Linux Calv-Lenovo 4.1.6-1-MANJARO #1 SMP PREEMPT Mon Aug 17 17:24:18 UTC 2015 x86_64 GNU/Linux
```

```
$ docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Sat Aug 15 17:29:10 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Sat Aug 15 17:29:10 UTC 2015
 OS/Arch:      linux/amd64
```

```
$ docker info
Containers: 1
Images: 51
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 53
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.6-1-MANJARO
Operating System: Manjaro Linux
CPUs: 8
Total Memory: 7.668 GiB
```
I first noticed the problem when my development DB container couldn't connect, and multiple containers after exhibited the same problem.

**Comment by mrjana** [comment]
@BenHall @calvcoll When this happens the next can you please provide the output of `iptables -L` and `iptables -t nat -L`?

**Comment by calvcoll** [comment]
Since mine seems to be ongoing, here's the outputs:
``` 
$ sudo iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain DOCKER (1 references)
target     prot opt source               destination         
ACCEPT     tcp  --  anywhere             172.17.0.4           tcp dpt:postgresql
```
```
$ sudo iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere            !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
MASQUERADE  all  --  172.17.0.0/16        anywhere            
MASQUERADE  tcp  --  172.17.0.4           172.17.0.4           tcp dpt:postgresql

Chain DOCKER (2 references)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             anywhere             tcp dpt:postgresql to:172.17.0.4:5432
```

**Comment by bprodoehl** [comment]
I've been having the same problem.  Same basic system stats as @BenHall 

```
# docker version
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   88cacbf
 Built:        Mon Aug 24 15:36:20 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   88cacbf
 Built:        Mon Aug 24 15:36:20 UTC 2015
 OS/Arch:      linux/amd64
```
```
# docker info
Containers: 136
Images: 119
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 391
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-62-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 1.947 GiB
Name: myhost
ID: 3ELI:V5B6:SQID:V3JS:KBN2:6PXF:JL6L:QENN:NHF6:PDDF:CWV4:3YU2
WARNING: No swap limit support
```
```
# uname -a
Linux myhost 3.13.0-62-generic #102-Ubuntu SMP Tue Aug 11 14:29:36 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
# 
```

And here are some containers trying to ping out:
```
# docker run ubuntu ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=56 time=0.610 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=56 time=0.530 ms
^C
--- 8.8.8.8 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.530/0.570/0.610/0.040 ms
# docker run ubuntu ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=56 time=0.777 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=56 time=0.730 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=56 time=0.462 ms
^C
--- 8.8.8.8 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2000ms
rtt min/avg/max/mdev = 0.462/0.656/0.777/0.140 ms
# docker run ubuntu ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
From 172.17.0.98 icmp_seq=1 Destination Host Unreachable
From 172.17.0.98 icmp_seq=2 Destination Host Unreachable
From 172.17.0.98 icmp_seq=3 Destination Host Unreachable
^C
--- 8.8.8.8 ping statistics ---
6 packets transmitted, 0 received, +3 errors, 100% packet loss, time 5031ms
pipe 3
# docker run ubuntu ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
From 172.17.0.99 icmp_seq=1 Destination Host Unreachable
From 172.17.0.99 icmp_seq=2 Destination Host Unreachable
From 172.17.0.99 icmp_seq=3 Destination Host Unreachable
^C
--- 8.8.8.8 ping statistics ---
5 packets transmitted, 0 received, +3 errors, 100% packet loss, time 4023ms
pipe 3
#
```

At the same time, the docker daemon is unable to reliably login to or pull images from the Docker Hub:
```
# docker pull ubuntu
Using default tag: latest
Pulling repository docker.io/library/ubuntu
Error while pulling image: Get https://index.docker.io/v1/repositories/library/ubuntu/images: dial tcp: lookup index.docker.io on 8.8.4.4:53: dial udp 8.8.4.4:53: network is unreachable
# 
```

**Comment by bprodoehl** [comment]
I was chasing down other problems with starting lots of containers through the API exhausting the system memory, causing odd leaks, and so far things look better when I run the daemon with --userland-proxy=false.  It'll take more testing to be sure that I don't hit this particular bad network state anymore, but I think it might be worth a try for anyone else encountering problems like this.

**Comment by BenHall** [comment]
@bprodoehl Very interesting, you may be right. We're starting everything via HTTP API and we're seeing the same results on a different server

**Comment by BenHall** [comment]
```
iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
MASQUERADE  all  --  172.17.0.0/16        anywhere
```

**Comment by bprodoehl** [comment]
@BenHall a solid night of beating on it with --userland-proxy=false revealed two new intermittent problems, neither of which I can find existing tickets for, so I'll create those now.  When you said you're seeing the same results on a different server, what exactly did you mean?

**Comment by bprodoehl** [comment]
I'm seeing this problem again on a machine with a master build from yesterday and --userland-proxy=false.

**Comment by bprodoehl** [comment]
@mavenugo I can provide access to this host as well, if you want to see it live.

**Comment by mavenugo** [comment]
@calavera we are working with @bprodoehl on https://github.com/docker/docker/issues/15341 which we think is the root-cause for all these issues.
@bprodoehl yes, that would certainly help.

**Comment by calavera** [comment]
cool, thanks.

**Comment by crosbymichael** [comment]
This has been fixed by #16038

thanks!
